feat: add guild member join/leave message

goestav opened the pull request:
This PR adds member join/leave messages.

Closes #4

goestav committed:
refactor: add example .env file

This makes it easier to set up and allows developers to simply copy paste the file.

goestav committed:
refactor: add newlines to main file

goestav committed:
feat: add guild member join/leave message

AsyncBanana commented on src/listeners/index.ts:
You might want to just use `export from` for brevity.

AsyncBanana commented on src/listeners/join-leave-message.ts:
```diff
@@ -0,0 +1,51 @@
+    })
+}
+
+const listener = (client: Client) => {
```
It seems more natural to either export the client from index.ts or another file and then import that directly, or export specific listeners that the client will handle (e.g exporting `handleGuildMemberJoin` as `guildMemberAdd`. It is up to you which to choose, just this approach seems necessarily verbose.

If we really wanted a simple API, globs would also be an option. However, at that point it would be better to implement this in djs-fsrouter.

AsyncBanana commented on .env.example:
```diff
@@ -3,3 +3,6 @@ TOKEN=
 GUILD=
+
+# Channel to send welcome/leave messages in
+GATEWAY_CHANNEL=
```
As you had already mentioned in Discord, we might want to implement this in the db for more flexibility. Now that the db is ready (we just need reviewers), if we do decide to use the db rather than an env value, you should wait for the db pr to be merged and then integrate that.

stifskere replied:
I would personally not add these variables in the ENV file and keep the ENV file for sensitive data like the TOKEN or the bot ID, and make an implementation later to be able to change these in runtime, maybe in the config database schema created in the other PR.

AsyncBanana reviewed (changes requested):
Overall, good job! Few potential changes to be made.

goestav committed:
refactor: use `export ... from ...` syntax

stifskere commented on src/listeners/join-leave-message.ts:
```diff
@@ -0,0 +1,51 @@
+    client.on('guildMemberRemove', handleGuildMemberLeave)
+}
+
+export default listener
```
This seems confusing for me, do we export the listener and then import it somewhere else? this just seems a more confusing implementation of just exporting functions and using `client.on('event', await import('./eventHandler'));`

stifskere commented on src/index.ts:
```diff
@@ -21,6 +24,17 @@ client.once(Events.ClientReady, async (bot) => {
+	} catch (err) {
+		console.error(`Error loading listeners!\n${err}`);
+	}
+
```
What about instead of wrapping the whole `for` loop inside a `try catch` block, just not wrap the addition?

```ts
		for (const listenerName in listeners) {
                        try {
			    listeners[listenerName as keyof typeof listeners](bot);
                        }
                        catch (e) {
                            console.error(`Error loading ${listenerName} due to ${e}`);
                        }
		}
```

And also standardize the usage of `;`, maybe.

goestav committed:
refactor: add basic editorconfig

goestav committed:
refactor: move try-catch block inside for-loop

goestav committed:
refactor: fix linter warning

goestav committed:
refactor: format code with rome

goestav committed:
refactor: improve event listener handler structure

goestav committed:
refactor: delete unused listeners main file

goestav committed:
refactor: switch to new listener structure

goestav committed:
refactor: use method definition syntax for event listeners

goestav commented:
Which syntax do you guys prefer for setting up event listeners? 🤔

Option 1: Helper function which can infer the type automatically (current implementation)
```ts
export default createListener({
    event: 'messageCreate',
    handler(message) {
        return message.reply('hi')
    }
})
```
Option 2: Plain object with argument type annotation
```ts
const listenerArgument: ListenerOptions = {
    event: 'messageCreate',
    handler(message: Message) {
        return message.reply('hi')
    }
}

export default listenerArgument
```
Option 3: Plain object with generic
```ts
const listenerGeneric: ListenerOptions<'messageCreate'> = {
    event: 'messageCreate',
    handler(message) {
        return message.reply('hi')
    }
}

export default listenerGeneric
```

([TS playground link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAqgzhATlAvFA3lARgQwHYBcUAFAJSoB8UAbgPYCWAJlAL4BQokUAshHHDgDm0NJkQQwAGxBFiAWz4DhROMET08g8iip0mrNh3DQAwpPoQ8wAKLVLwOKgxsorqAHIF-IRBPicwBDuRADavN7CALoubu4ArgiIfhABQaHwSNHsRlwAMvSqlkgAKsZOAES0eOVQAD5QlXgAxhDlANyGGoGIAGY4LVD5hXhIAPJgwPRVcAA8bACQxVAQAB6BeIyOANYQILQ9UGYWVrb2jmg7ewdH9qdWcAA0bFToCwD0AFQfUAACjBB9OKSYANKo1D5vBacCAAfiIQ3WJWMHXmEDsViIxRRAAt8IxJEhZAA6Ek4RCCOBEG4ndEOELFSLaXQMZj1AAKiFocgKEBmekYFA62WhgwKiMQMyWq3Wmygl32h3Mt1p5zluwV1JsKqoKAWACUIABHOL0cSMGYIoqIcaTaaSiiCwxNaYgpr+QKWkbINBzRbLNaWWXy65KmlnJzBxXHLVnJ4UYgLczDQmi5PWiZTPCzYoUJ6keFiq32ygkV7zaFERrlJ7zElEpPip4sUgdQxsT5QUZs4oASVGADkoABGIgACQgkkgyB6cWatrwUAA7tj6E1sVAmvgoBoekgoMBsdARTg4sAuQFVzhJNISE04ohxFZt3IpBAFFYL1VyBCnS6oA2rScN0Ug9QsvWIV43GWWkiE8RQfGSVJ3CeKDcQ2AlEHkeDhHISCoLccRgHvBcvCUCAiXEKQQGIdwV3cUgYlcdhm0MDsu17AcoAAJiINlJBwDQoFoLAACsICaEFF3oA8oDJQQ4nfEFjzwPBaGAT8Fx-NhnSzEEAK9ABBckFPsAs0xtTNVTw1w0VMjxSIQ90ghQtw3jeX4HAAWlWSAJO8h9aGQegDkXaB-iaJgj1oKAEmgKQBIXYSxIkxwD0sfdD23KwkD6AYwE5KdQH3MpF1oIFmCwaBZy2VTFwXOBooPAolzKyRGDwdwQSQTlEEYqA0PxJAsIiCAiHCMjcL6qDCOIqAHOECiJGkGi6IYqDmNY752L7QcAGZeP4wSkvEyTpPXYQvVXKAtJ01R-zApAAHErVXMzxQsu04JGxDAncHVnCg2yMXs7DfCc5C+oGjDhomgH8NcGbEBI0HFqolb6HovrmKAA))

AsyncBanana commented:
Is it possible to use unions to make the generic unnecessary? I think you could create a union of the possible events with their `event` property set to a specific value and their handler set with the correct param so the `event` property defines the `handler` type.

However, if that is impossible or too hard/tedious to implement, I think option 3 is best.

goestav commented:
It gets inferred as a union when you don't annotate it with a type but then it has problems with reading the properties/methods
And we'd have to cast it to the right type inside the function like `return (message as Message).reply('hi')`
<img width="608" alt="image" src="https://github.com/Javascripters-Development/JavaScripters-Bot/assets/27970303/33a249bc-d598-4069-a214-8806fdedfcf1">

Edit: I guess if we turn it into an intersection that it would work 🤔 but then we also have *all* properties and methods available which is kind of useless then

GlanVonBrylan commented on src/index.ts:
```diff
@@ -1,5 +1,9 @@
 import loadCommands from "djs-fsrouter";
 import { Client, GatewayIntentBits, Events } from "discord.js";
+import { join } from "path";
+import { readdir } from "fs/promises";
```
```suggestion
import { readdir } from "fs/promises";
```

goestav committed:
refactor: merge `gatewayEnabled` with `gatewayChannel` column

If the `gatewayChannel` is NULL then we can assume the gateway is disabled, otherwise enabled.

AsyncBanana commented:
You should be able to attach a specific `type` value to each type in the union to narrow down parameter types like so:
```ts
interface messageCreate {
   event: "messageCreate";
   handler: (message: string) => void;
}
interface guildMemberAdd {
    event: "guildMemberAdd";
    handler: (user: number) => void;
}
type DiscordEventListener = messageCreate | guildMemberAdd;
```

TypeScript should be able to figure out whether the object is messageCreate or guildMemberAdd from the `type` property, and then apply their respective types to the rest of the properties.

In fact, if you wanted you could probably build the union out of generics like the one above, and simply set `type` to the argument passed to the generic (with setting the parameter types of the handler)

goestav committed:
refactor: use plain objects for event listeners

goestav committed:
refactor: construct embed from database

goestav committed:
refactor: fetch gateway channel from database

AsyncBanana commented on .editorconfig:
```diff
@@ -0,0 +1,9 @@
+
+[*.ts,*.json]
+indent_style = space
+indent_size = 4
```
I forgot to add that we should remove this to avoid confusion on how to format the code.

goestav replied:
removed it

AsyncBanana commented on src/listeners/join-leave-message.ts:
```diff
@@ -0,0 +1,76 @@
+	const [firstRow] = await db
+		.select({ gatewayChannel: Config.gatewayChannel })
+		.from(Config)
+		.limit(1);
```
This query looks like its missing a .where() clause to search for the correct guild id. Also, with the where clause, limit will be unecessary.

It is also a good idea to use [prepared statements](https://orm.drizzle.team/docs/performance) with placeholders for the guild id, as this will help performance quite a bit.

Finally, you can use `.get` to automatically return the first row so you don't need to use destructuring.

goestav replied:
I can't figure out how to do that properly, I'm doing it like it says in the docs.
It tells me that the `placeholder` method is deprecated (with the message "Use `sql.placeholder` instead.") but the usage of that isn't mentioned in the docs anywhere.

AsyncBanana replied:
As far as I can tell, the API for `sql.placeholder` is exactly the same as `placeholder`. Just `import {sql} from "drizzle-orm"` and use that.

AsyncBanana commented on src/listeners/join-leave-message.ts:
```diff
@@ -0,0 +1,76 @@
+				: Config.gatewayJoinContent,
+		})
+		.from(Config)
+		.limit(1);
```
This should probably be merged into the above query, as both are operating on the same row.

AsyncBanana commented on src/listener.ts:
```diff
@@ -0,0 +1,15 @@
+	};
+};
+
+export type Listener = MappedListeners[keyof MappedListeners];
```
This would probably be better off in a `/types` directory considering this has no runtime functionality. Also, in general, try to remember to add type to the imports, e.g. `import type { Awaitable } from "discord.js"

goestav replied:
I'll enable [verbatimModuleSyntax](https://www.typescriptlang.org/tsconfig/#verbatimModuleSyntax) in the tsconfig

AsyncBanana reviewed (changes requested):
There are a few more smaller things, primarily relating to the database queries.

goestav committed:
refactor: delete editorconfig

goestav committed:
refactor: use where clause

goestav committed:
refactor: use `.get()` to retrieve a single row

goestav committed:
refactor: move listener types to types directory

goestav committed:
build: require `import type` for type-only imports

goestav committed:
refactor: use single query to get the database config

goestav committed:
refactor: add utility function for retrieving the guild config

AsyncBanana commented on src/utils.ts:
```diff
@@ -0,0 +1,32 @@
+		.from(Config)
+		.where(eq(Config.id, sql.placeholder("guildId")))
+		.prepare()
+		.get({ guildId });
```
You need to prepare the statement at startup and then run the prepared statement. Preparing it every time the query is executed removes the point of a prepared statement.

goestav committed:
refactor: export prepared statement from utils

AsyncBanana committed:
refactor: reorganize code
